src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,198 @@
+import {
+    AnyImportOrRequireStatement,
+    append,
+    ApplicableRefactorInfo,
+    codefix,
+    Debug,
+    Diagnostics,
+    emptyArray,
+    getLocaleSpecificMessage,
+    getQuotePreference,
+    hasSyntacticModifier,
+    hostGetCanonicalFileName,
+    Identifier,
+    insertImports,
+    isPrologueDirective,
+    LanguageServiceHost,
+    ModifierFlags,
+    nodeSeenTracker,
+    Program,
+    QuotePreference,
+    RefactorContext,
+    RefactorEditInfo,
+    skipAlias,
+    SourceFile,
+    Symbol,
+    SyntaxKind,
+    takeWhile,
+    textChanges,
+    TypeChecker,
+    UserPreferences,
+} from "../_namespaces/ts";
+import { addExports, addExportToChanges, addNewFileToTsconfig, createOldFileImportsFromNewFile, deleteMovedStatements, deleteUnusedOldImports, filterImport, forEachImportInStatement, getStatementsToMove, getTopLevelDeclarationStatement, getUsageInfo, isTopLevelDeclaration, makeImportOrRequire, moduleSpecifierFromImport, nameOfTopLevelDeclaration, ToMove, updateImportsInOtherFiles, UsageInfo } from "../moveToFileAndNewFile";
+import { registerRefactor } from "../refactorProvider";
+
+const refactorNameForMoveToFile = "Move to file";
+const description = getLocaleSpecificMessage(Diagnostics.Move_to_file);
+
+const moveToFileAction = {
+    name: "Move to file",
+    description,
+    kind: "refactor.move.file",
+};
+registerRefactor(refactorNameForMoveToFile, {
+    kinds: [moveToFileAction.kind],
+    getAvailableActions: function getRefactorActionsToMoveToFile(context): readonly ApplicableRefactorInfo[] {
+        const statements = getStatementsToMove(context);
+        if (context.preferences.allowTextChangesInNewFiles && statements) {
+            return [{ name: refactorNameForMoveToFile, description, actions: [moveToFileAction] }];
+        }
+        if (context.preferences.provideRefactorNotApplicableReason) {
+            return [{ name: refactorNameForMoveToFile, description, actions:
+                [{ ...moveToFileAction, notApplicableReason: getLocaleSpecificMessage(Diagnostics.Selection_is_not_a_valid_statement_or_statements) }]
+            }];
+        }
+        return emptyArray;
+    },
+    getEditsForAction: function getRefactorEditsToMoveToFile(context, actionName, newFile): RefactorEditInfo | undefined {
+        Debug.assert(actionName === refactorNameForMoveToFile, "Wrong refactor invoked");
+        const statements = Debug.checkDefined(getStatementsToMove(context));
+        if (newFile) {
+            const edits = textChanges.ChangeTracker.with(context, t => doChange(context, context.file, newFile, context.program, statements, t, context.host, context.preferences));
+            return { edits, renameFilename: undefined, renameLocation: undefined };
+        }
+        return undefined;

```

Comment:
Discussed in chat, and this is related to the API questions I posed earlier, but I think it would be better to assert `newFile` is defined, because if it’s not, the caller is definitely using the API wrong, and getting a useful error message is probably better than getting nothing.